jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentCopier.java

```diff
@@ -46,43 +47,41 @@ public InvocationType getInvocationType()
     @Override
     protected Action process() throws Throwable
     {
+        if (current != null)
+            current.release();
```

Comment:
I don't like mixing having methods overriding raw methods (`succeeded()`, `failed(Throwable)` and `abort(Throwable)`) with the processed methods  (`onXxx(...)`).  I think it is too easy to end up writing code that is very dependent on the internal implementation between raw and processed.  Hence I prefer it here.

Reply:
It's weird that it is at the top of the iteration.
One reads the code, and goes "why do we need to release this first thing, and why the null check?" to realize that is necessary _only_ when `succeeded()` is called (and no other method), so it really feels that `succeeded()` is the right way to put this (and the null check is only necessary for other cases, as it would not be necessary in `succeeded()`).
In fact, succeeded() is the right place to put any "cleanup" code that needs to be executed per successful iteration.
Having it in the loop is like having a for loop, but with the increment removed from the last expression of the for statement and put into the loop -- works but it's weird :)

Reply:
@sbordet, I think succeeded is precisely the wrong place to do anything... I'm tempted to make it a final method.    
 We go to a lot of effort to serialize threads so that only a single thread is in `process()` at any given time.  The whole point of `process()` is that it is a pumping method to move the state on with well defined concurrency.   It is natural that it may reuse and/or clean up from previous iterations.

Doing anything in `succeeded()` has no defined concurrency and we could be changing fields that are actively in use by a processing thread.  For this particular usage, it is probably not a problem, but the following variation could throw a NPE with debug turned on if succeeded does the releasing:
```
        sink.write(current.isLast(), current.getByteBuffer(), this);
        Log.debug("writing {}", current.isLast());
        return Action.SCHEDULED;
```

We should declare it an anti-pattern for any class to extend both a `succeeded()` and an `onXxx()` method.